include/translated-object.php

```diff
@@ -3,176 +3,113 @@
+		parent::__construct( $model );
+
+		$this->tax_to_cache[] = $this->tax_translations;
 
 		/*
-		 * Register our taxonomies as soon as possible.
-		 * This is early registration, not ready for rewrite rules as $wp_rewrite will be setup later.
+		 * Register our taxonomy as soon as possible.
 		 */
-		$args = array( 'label' => false, 'public' => false, 'query_var' => false, 'rewrite' => false, '_pll' => true );
-		register_taxonomy( $this->tax_language, $this->object_type, $args );
-		$args['update_count_callback'] = '_update_generic_term_count'; // Count *all* objects to avoid deleting in clean_translations_terms.
-		register_taxonomy( $this->tax_translations, $this->object_type, $args );
+		register_taxonomy(
+			$this->tax_translations,
+			(array) $this->object_type,
+			array(
+				'label'                 => false,
+				'public'                => false,
+				'query_var'             => false,
+				'rewrite'               => false,
+				'_pll'                  => true,
+				'update_count_callback' => '_update_generic_term_count', // Count *all* objects to avoid deleting in clean_translations_terms.
+			)
+		);
 	}
 
 	/**
-	 * Stores the language in the database.
-	 *
-	 * @since 0.6
+	 * Returns the translations group taxonomy name.
 	 *
-	 * @param int                     $id   Object id.
-	 * @param int|string|PLL_Language $lang Language (term_id or slug or object).
-	 * @return void
-	 */
-	abstract public function set_language( $id, $lang );
-
-	/**
-	 * Returns the language of an object.
+	 * @since 3.4
 	 *
-	 * @since 0.1
+	 * @return string
 	 *
-	 * @param int $id Object id.
-	 * @return PLL_Language|false PLL_Language object, false if no language is associated to that object.
+	 * @phpstan-return non-empty-string
 	 */
-	abstract public function get_language( $id );
+	public function get_tax_translations() {
+		return $this->tax_translations;
+	}
 
 	/**
 	 * Assigns a new language to an object, taking care of the translations group.
 	 *
 	 * @since 3.1
 	 *
-	 * @param int          $id   Object id.
+	 * @param int          $id   Object ID.
 	 * @param PLL_Language $lang New language to assign to the object.
-	 * @return void
+	 * @return bool True when successfully assigned. False otherwise (or if the given language is already assigned to
+	 *              the object).
 	 */
-	public function update_language( $id, $lang ) {
-		$id = $this->sanitize_int_id( $id );
-
-		if ( empty( $id ) || $this->get_language( $id ) === $lang ) {
-			return;
+	public function update_language( $id, PLL_Language $lang ) {
+		if ( ! $this->set_language( $id, $lang ) ) {
+			return false;
 		}
 
-		$this->set_language( $id, $lang );
+		$id = $this->sanitize_int_id( $id );
 
 		$translations = $this->get_translations( $id );
 
-		if ( $translations ) {
-			// Remove the post's former language from the new translations group.
+		// Don't create translation groups with only 1 value.
+		if ( ! empty( $translations ) ) {
```

Comment:
If I'm not wring, term translations are stored even if they contains one value for the moment.
This would avoid that? Could it be a problem @Chouby?